check/src/unify_type.rs

```diff
@@ -171,9 +171,11 @@ fn do_zip_match<'a, U>(unifier: &mut UnifierState<'a, U>,
 {
     debug!("Unifying:\n{:?} <=> {:?}", expected, actual);
     match (&**expected, &**actual) {
-        (&Type::App(ref l, ref l_args), &Type::App(ref r, ref r_args)) => {
-            Ok(unify_app(unifier, l, l_args, r, r_args))
+        (&Type::App(ref l_typ, ref l_args), &Type::App(ref r_typ, ref r_args)) => {
+            Ok(unify_app(unifier, l_typ, l_args, r_typ, r_args))
         }
+        (&Type::Forall(_, ref l_typ), _) => Ok(unifier.try_match(l_typ, actual)),
+        (_, &Type::Forall(_, ref r_typ)) => Ok(unifier.try_match(expected, r_typ)),
```

Comment:
It might not be unsound but it may give weird results since the types will contain uninstantiated variables (`Type::Generic(_)`).

If you do instantiate the forall then I think it needs to be only `forall ... <=> forall ...` or else this may unify which I don't believe is sound
`forall b. { id : forall a . a -> a, x: b } <=> forall a b. { id: a -> a, x: b}`